File: src/main/java/org/hisp/dhis/rules/models/RuleActionDisplayKeyValuePair.java
package org.hisp.dhis.rules.models;

import com.google.auto.value.AutoValue;

import javax.annotation.Nonnull;
import javax.annotation.CheckForNull;

@AutoValue
public abstract class RuleActionDisplayKeyValuePair
    extends RuleActionText
{

    @Nonnull
    public static RuleActionDisplayKeyValuePair createForFeedback(
        @CheckForNull String content, @CheckForNull String data )
    {
        if ( content == null && data == null )
        {
            throw new IllegalArgumentException( "Both content and data must not be null" );
        }

        return new AutoValue_RuleActionDisplayKeyValuePair( data == null ? "" : data, content == null ? "" : content,
            LOCATION_FEEDBACK_WIDGET );
    }

    @Nonnull
    public static RuleActionDisplayKeyValuePair createForIndicators(
        @CheckForNull String content, @CheckForNull String data )
    {
        if ( content == null && data == null )
        {
            throw new IllegalArgumentException( "Both content and data must not be null" );
        }

        return new AutoValue_RuleActionDisplayKeyValuePair( data == null ? "" : data, content == null ? "" : content,
            LOCATION_INDICATOR_WIDGET );
    }
}

File: src/main/java/org/hisp/dhis/rules/models/RuleActionDisplayKeyValuePair.kt
package org.hisp.dhis.rules.models

import javax.annotation.concurrent.Immutable

@Immutable
data class RuleActionDisplayKeyValuePair(override var content: String?, override var data: String?, override var location: String?) : RuleActionText() {
    companion object {

        fun createForFeedback(
                content: String?, data: String?): RuleActionDisplayKeyValuePair {
            if (content == null && data == null) {
                throw IllegalArgumentException("Both content and data must not be null")
            }

            return RuleActionDisplayKeyValuePair(content ?: "",
                    data ?: "", RuleActionText.LOCATION_FEEDBACK_WIDGET)
        }

        fun createForIndicators(
                content: String?, data: String?): RuleActionDisplayKeyValuePair {
            if (content == null && data == null) {
                throw IllegalArgumentException("Both content and data must not be null")
            }

            return RuleActionDisplayKeyValuePair(content ?: "",
                    data ?: "", RuleActionText.LOCATION_INDICATOR_WIDGET)
        }
    }
}
